File: src/backend/dotNet/dotNet/Models/ANNSettings.cs


namespace dotNet.Models {
    public class ANNSettings {

        // General info
        public ProblemType ANNType { get; set; }

        public ANNSettings(
            ProblemType aNNType, 
            float learningRate, 
            int batchSize,
            int numberOfEpochs, 
            int currentEpoch, 
            int inputSize, 
            int outputSize, 
            int[] hiddenLayers, 
            ActivationFunction[]? activationFunctions,
            RegularizationMethod regularization,
            float regularizationRate,
            LossFunction lossFunction, 
            Optimizer optimizer,
            float[] optimizationParams,
            int kFoldCV) {

            ANNType = aNNType;
            LearningRate = learningRate;
            BatchSize = batchSize;
            NumberOfEpochs = numberOfEpochs;
            CurrentEpoch = currentEpoch;
            InputSize = inputSize;
            OutputSize = outputSize;
            HiddenLayers = hiddenLayers;
            ActivationFunctions = activationFunctions;
            Regularization = regularization;
            RegularizationRate = regularizationRate;
            LossFunction = lossFunction;
            Optimizer = optimizer;
            OptimizationParams = optimizationParams;
            KFoldCV = kFoldCV;
        }

        public float LearningRate { get; set;}
        public int BatchSize { get; set; }
        public int NumberOfEpochs { get; set; }
        public int CurrentEpoch { get; set; }
        // IO
        public int InputSize { get; set; }
        public int OutputSize { get; set; }
        // Layers
        public int[]? HiddenLayers { get; set; }
        public ActivationFunction[]? ActivationFunctions { get; set; }
        // Other
        public RegularizationMethod Regularization { get; set; }
        public float RegularizationRate { get; set; }
        public LossFunction LossFunction { get; set; }
        public Optimizer Optimizer { get; set; }
        public float[]? OptimizationParams { get; set; }
        public int KFoldCV { get; set; }

        // To JSON string
        public override string ToString() {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this);
        }
    }

    public enum ProblemType {
        Regression,
        Classification
    }

    public enum ActivationFunction {
        ReLU,
        LeakyReLU,
        Sigmoid,
        Tanh,
        Linear
    }

    public enum RegularizationMethod {
        L1,
        L2
    }

    public enum LossFunction {
        // Regression
        L1Loss,
        L2Loss,
        SmoothL1Loss,
        HuberLoss,
        // Classification
        NLLLoss,
        CrossEntropyLoss,
        KLDivLoss,
        MultiMarginLoss
    }

    public enum Optimizer {
        Adadelta,
        Adagrad,
        Adam,
        AdamW,
        Adamax,
        ASGD,
        NAdam,
        RAdam,
        RMSprop,
        SGD
    }
}

File: src/ml/Models/ANNSettings.py

import json


class ANNSettings:
    
    
    def __init__(
        self,
        problemType         = 0, 
        learningRate        = 0.0, 
        batchSize           = 0, 
        numberOfEpochs      = 0, 
        currentEpoch        = 0,
        inputSize           = 0, 
        outputSize          = 0, 
        hiddenLayers        = None, 
        activationFunctions = None,
        regularization      = 0,
        regularizationRate  = 0.0,
        lossFunction        = 0,
        optimizer           = 0,
        optimizationParams  = [0.0],
        kFoldCV             = 0
        ) -> None:
        
        self.problemType         = problemType
        self.learningRate        = learningRate
        self.batchSize           = batchSize
        self.currentEpoch        = currentEpoch
        self.numberOfEpochs      = numberOfEpochs
        self.inputSize           = inputSize
        self.outputSize          = outputSize
        self.hiddenLayers        = hiddenLayers
        self.activationFunctions = activationFunctions
        self.regularization      = regularization
        self.regularizationRate  = regularizationRate
        self.lossFunction        = lossFunction
        self.optimizer           = optimizer
        self.optimizationParams  = optimizationParams
        self.kFoldCV             = kFoldCV
    
    def load(data) -> None:
        jsonObj = json.loads(data)
        return ANNSettings(
            problemType         = jsonObj["ANNType"],
            learningRate        = jsonObj["LearningRate"],
            batchSize           = jsonObj["BatchSize"],
            numberOfEpochs      = jsonObj["NumberOfEpochs"],
            currentEpoch        = jsonObj["CurrentEpoch"],
            inputSize           = jsonObj["InputSize"],
            outputSize          = jsonObj["OutputSize"],
            hiddenLayers        = jsonObj["HiddenLayers"],
            activationFunctions = jsonObj["ActivationFunctions"],
            regularization      = jsonObj["Regularization"],
            regularizationRate  = jsonObj["RegularizationRate"],
            lossFunction        = jsonObj["LossFunction"],
            optimizer           = jsonObj["Optimizer"],
            optimizationParams  = jsonObj["OptimizationParams"],
            kFoldCV             = jsonObj["KFoldCV"]
        )
    
    
    